src/app-layer.c

```diff
@@ -631,13 +633,21 @@ int AppLayerHandleTCPData(ThreadVars *tv, TcpReassemblyThreadCtx *ra_ctx,
     } else if (alproto != ALPROTO_UNKNOWN && FlowChangeProto(f)) {
         f->alproto_orig = f->alproto;
         SCLogDebug("protocol change, old %s", AppProtoToString(f->alproto_orig));
+        void *alstate_orig = f->alstate;
+        AppLayerParserState *alparser = f->alparser;
+        // we delay AppLayerParserStateCleanup because we may need previous parser state
         AppLayerProtoDetectReset(f);
         /* rerun protocol detection */
         if (TCPProtoDetect(tv, ra_ctx, app_tctx, p, f, ssn, stream,
                            data, data_len, flags) != 0) {
             SCLogDebug("proto detect failure");
+            AppLayerParserStateProtoCleanup(f->protomap, f->alproto_orig, alstate_orig, alparser);
             goto failure;
         }
+        if (f->alproto_orig == ALPROTO_HTTP && f->alproto_expect == ALPROTO_HTTP2) {
```

Comment:
the goal should be to get rid of this and instead move this to the http2 implementation, which should then be able to access the old http1 state to do its initial setup

Reply:
Ok, so adding a field `alstate_orig` to the Flow structure ?

Reply:
Thats the easiest way, but I'm not completely happy with it. It means we'll spend 8 extra bytes for each extra flow. I wonder if we can do something like have a special flag that indicates that we're going to perform an upgrade, so `f->alstate` contains the old state, which will be replaced by the new state in `AppLayerParserParse`. Since we need it only at one moment, perhaps we can expose the old state to the new parsers state creation function via a simple thread local variable. A bit hacky as well though.

Reply:
> It means we'll spend 8 extra bytes for each extra flow.

Indeed, and this current PR does not need those extra memory consumption

> which will be replaced by the new state in AppLayerParserParse

We need this new state in `HTTP2MimicHttp1Request`

So, if this is to be called from http2 implementation, this can only be in its `AppLayerParserParse`
So, we have alstate, pstate, and `alp_tctx->alproto_local_storage[f->protomap][alproto]` available
Is this last one your suggestion so that HTTP2 parser when checking a flag, we access there somehow the HTTP1 state

I am not sure it is better in the HTTP2 implementation, than in this condition with `FlowChangeProto`

Another solution is to call a new generic function `AppLayerParserProtoChange(f, alstate_orig)` and make it call some new parser's defined function `ProtoChange`